### PR TITLE
add nelder_mead

### PR DESCRIPTION
**Pull request by joaoluisro**
I'm really unsure about 
``` julia
  vertices = hcat(vertices, map(x->obj(nlp,x), vertices))
  vertices = vertices[sortperm(vertices[:, 2]), :]
```
to concatenate both vertices points and its corresponding objective values.
Would love to hear some suggestions and feedback.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -8,6 +8,7 @@ LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 NLPModels = "a4795742-8479-5a88-8948-cc11e1c8c1a6"
 Printf = "de0858da-6303-5e67-8744-51eddeeeb8d7"
 Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
+Revise = "295af30f-e4ad-537b-8983-00126c2a3abe"
```
Comment (abelsiqueira) [inline Project.toml]: Remove this

File: src/DerivativeFreeSolvers.jl [hunk 2]
```diff
@@ -8,5 +8,5 @@ using LinearAlgebra, Printf
 
 include("coordinate_search.jl")
 include("mads.jl")
-
+include("nelder_mead.jl")
```
Comment (abelsiqueira) [inline src/DerivativeFreeSolvers.jl:11]: Leave the space before the `end`

File: src/nelder_mead.jl [hunk 3]
```diff
@@ -0,0 +1,113 @@
+export nelder_mead
```
Comment (abelsiqueira) [inline src/nelder_mead.jl:1]: Add line

File: src/nelder_mead.jl [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+"""
+    nelder_mead(nlp)
+
+This implementation follows the algorithm described in chapter 9 of [1].
+[1] Numerical Optimization (Jorge Nocedal and Stephen J. Wright), Springer, 2006.
+"""
+function nelder_mead(nlp :: AbstractNLPModel;
+                     x :: AbstractVector = copy(nlp.meta.x0),
+                     vertices :: AbstractArray = Array{typeof(x)}(undef, nlp.meta.nvar+1),
```
Comment (abelsiqueira) [inline src/nelder_mead.jl]: Change to
```
vertices :: Vector{<: AbstractVector{<: AbstractFloat}} = Vector{eltype(x)}[]
```
so it's initially an empty vector of vectors. We'll have to keep an eye on type stability in the near future.

File: src/nelder_mead.jl [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                     max_eval :: Int = -1,
+                     ref_prm :: Real = - one(eltype(x)),
+                     exp_prm :: Real = - one(eltype(x)) * 2,
+                     ocn_prm :: Real = - one(eltype(x)) / 2,
+                     icn_prm :: Real = one(eltype(x)) / 2)
+
+  # Initial simplex
+  n = nlp.meta.nvar
+  T = eltype(x)
+  if !(isassigned(vertices))
```
Comment (abelsiqueira) [inline src/nelder_mead.jl]: Compute `length(vertices)`. If it's not 0 or n+1, then raise an error. If is 0, then create the memory for them all with
```
vertices = [x, [copy(x) for i = 0:n]...]
```

File: src/nelder_mead.jl [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    vertices[1] = x
+    for j in 1:n
+      xt = copy(x)
+      xt[j] += xt[j] == 0 ? eps(T)^T(0.25) : T(0.05)
+      vertices[j+1] = xt
```
Comment (abelsiqueira) [inline src/nelder_mead.jl]: Inside the `if`:
```
for j = 1:n
  xt = vertices[j + 1]
  xt[j] += ...
end
```
`xt = ...` creates a pointer, so it's all good. And memory was created before.

File: src/nelder_mead.jl [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    end
+  end
+
+  vertices = hcat(vertices, map(x->obj(nlp, x), vertices))
+  vertices = vertices[sortperm(vertices[:, 2]), :]
```
Comment (abelsiqueira) [inline src/nelder_mead.jl]: Create a new structure to store these pairs, because of type stability.
```
pairs = [(x, obj(nlp, x)) for x in vertices]
```

Comment (abelsiqueira) [reply src/nelder_mead.jl]: And sort
```
sort!(pairs, by=x->x[2])
```

File: src/nelder_mead.jl [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    μ(t) = Array{T,1}(x_cen + t * (vertices[n+1, 1] - x_cen))
+    x_ref = μ(ref_prm)
+    f_ref = obj(nlp, x_ref)
+    f_bver = vertices[1, 2]
+
+    @info log_row(Any[k, f_bver, norm_opt])
+
+    # Reflection
+    if vertices[1, 2] ≤ f_ref < vertices[n, 2]
+      vertices[n+1, 1] = x_ref
```
Comment (abelsiqueira) [inline src/nelder_mead.jl]: All these assignments need two things:
- Reuse memory. Outside the while, create a `xt` which will store all trial points. I believe `μ(t)` will have to be removed in favor of reusing the memory. Then assign with `.=` to reuse the memory.
- Reuse the objective value. Assume the largest cost inside the algorithm is computing `obj`, and always reuse when possible. (See https://bbcomp.ini.rub.de/)

File: src/nelder_mead.jl [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        if f_icn < vertices[n+1, 2]
+          vertices[n+1, 1]= x_icn
+          shrink = false
+        end
+      end
+    end
+
+    if shrink
+      vertices[2:n+1, 1] = map(u -> (u + vertices[1, 1])/2, vertices[2:n+1, 1])
+    end
```
Comment (abelsiqueira) [inline src/nelder_mead.jl:106]: Change to a `for`, reuse memory, and compute the objective here (which will be the first time for each new point.)

File: src/nelder_mead.jl [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+    vertices[1:n+1, 2] = map(u->obj(nlp, u), vertices[1:n+1, 1])
```
Comment (abelsiqueira) [inline src/nelder_mead.jl]: Remove

File: src/nelder_mead.jl [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    vertices = vertices[sortperm(vertices[:, 2]), :]
```
Comment (abelsiqueira) [inline src/nelder_mead.jl]: `sort!`

**Review by abelsiqueira — CHANGES_REQUESTED** [review]
Thanks for the PR. I've made some changes to improve the code. I think you'll like the new commands.

File: src/nelder_mead.jl (hunk 8 above)
Comment (joaoluisro) [reply src/nelder_mead.jl]: You mean allocating ```xt``` as
```julia
   xt = [zeros(n) for i = 1:4]
```
and using it like :
```julia
    x_cen = sum(pairs[i][1] for i in 1:n)/T(n)
    # xt[1] := reflection point instead of x_ref
    xt[1] .= x_cen * (1 - ref_prm) + ref_prm * pairs[n+1][1]
    f_ref = obj(nlp, xt[1])

    # Reflection
    if pairs[1][2] ≤ f_ref < pairs[n][2]
      pairs[n+1][1] = (xt[1], f_ref)
```
 If so, i feel it's a little awkward to keep track of each index of every parameter. Isn't there a better way to both reuse memory and keep clean legibility, or it isn't such a big deal ? Just a thought, i'm sorry if that's not what you meant.

File: src/nelder_mead.jl (hunk 5 above)
Comment (joaoluisro) [reply src/nelder_mead.jl]: Shouldn't it be ```vertices = [copy(x) for i = 0:n]```  ? Since it's copying the ```x``` vector ```n+1``` times. Or ```vertices = [x, [copy(x) for i = 1:n]...]```

Comment (abelsiqueira) [reply src/nelder_mead.jl]: It's `i = 1:n`, thanks.

File: src/nelder_mead.jl [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+  norm_opt = norm(first(pairs)[1] - last(pairs)[1])
+  optimal = norm_opt < tol
+  T = eltype(x)
+  status =:unknown
+  @info log_header([:iter, :f, :nrm], [Int, T, T], hdr_override=Dict(:f => "f(x)", :nrm => "‖x₁ - xₙ+1‖"))
+
+  while !(optimal || tired)
+
+    shrink = true
+    x_cen = sum(pairs[i][1] for i in 1:n)/T(n)
```
Comment (joaoluisro) [inline src/nelder_mead.jl]: The ``` .= ``` assignment here (to compute the reflection point) seems to break the algorithm somehow, i'll have to take a deeper look at it.

Comment (abelsiqueira) [reply src/nelder_mead.jl]: `.=` only works if the lhs is defined, I think `x_cen` is first defined here? Create it outside the loop. (`zeros(T, n)`).

**Comment by joaoluisro** [comment]
I've implemented the requested changes, using tuples seems to make a lot more sense.

File: src/nelder_mead.jl [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    sort!(pairs, by=x->x[2])
+    k += 1
+    tired = neval_obj(nlp) > max_eval ≥ 0 || el_time > max_time
+    norm_opt = norm(first(pairs)[1] - last(pairs)[1])
+    optimal =  norm_opt < tol
+    el_time = time() - start_time
+  end
+
+  if optimal
+    status = :unknown
```
Comment (abelsiqueira) [inline src/nelder_mead.jl]: `status = :acceptable`

Comment (abelsiqueira) [reply src/nelder_mead.jl]: https://github.com/JuliaSmoothOptimizers/SolverTools.jl/blob/master/src/stats/stats.jl#L7

**Comment by joaoluisro** [comment]
Everything set.

**Comment by joaoluisro** [comment]
Everything should be in order.

File: src/nelder_mead.jl [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+          pairs[n+1][1] .= x_trial
+          pairs[n+1][2] = f_icn
+          shrink = false
+        end
+      end
+    end
+
+    if shrink
+      for i = 2:n+1
+        aux = (pairs[i][1] + pairs[1][1])/2
```
Comment (abelsiqueira) [inline src/nelder_mead.jl]: You can reuse the memory of this `aux` too. Use `x_trial` here.

File: src/nelder_mead.jl [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+    tired = neval_obj(nlp) > max_eval ≥ 0 || el_time > max_time
+    norm_opt = norm(first(pairs)[1] - last(pairs)[1])
+    optimal =  norm_opt < tol
+    el_time = time() - start_time
+  end
+
+  if optimal
+    status = :acceptable
+  elseif tired
+    if neval_obj(nlp) > max_eval ≥  0
```
Comment (abelsiqueira) [inline src/nelder_mead.jl]: Remove extra space before 0

**Comment by joaoluisro** [comment]
Now everything should definitely be in order.

**Comment by abelsiqueira** [comment]
> Now everything should definitely be in order.

Famous last words.

File: Manifest.toml [hunk 16]
```diff
@@ -0,0 +1,299 @@
+# This file is machine-generated - editing it directly is not advised
```
Comment (abelsiqueira) [inline Manifest.toml]: This file should not be included in the commit. Probably the reason for the error.

File: test/Manifest.toml [hunk 17]
```diff
@@ -0,0 +1,204 @@
+# This file is machine-generated - editing it directly is not advised
```
Comment (abelsiqueira) [inline test/Manifest.toml]: This file should no be here either.
I recommend creating a global gitignore and adding `Manifest.toml` to it.

File: test/Project.toml [hunk 18]
```diff
@@ -0,0 +1,4 @@
+[deps]
```
Comment (abelsiqueira) [inline test/Project.toml]: `test` should not have a `Project.toml`, instead the main `Project.toml` has a section with the `Test` libraries.

Comment (abelsiqueira) [reply test/Project.toml]: Here's an example: https://github.com/JuliaSmoothOptimizers/SolverTools.jl/blob/master/Project.toml

**Comment by joaoluisro** [comment]
> > Now everything should definitely be in order.
> 
> Famous last words.

Not gonna lie, i accidentally did ```git add --all``` by mistake, my bad.

**Comment by joaoluisro** [comment]
Fingers crossed that this one seals the deal.

File: src/nelder_mead.jl [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+                     max_eval :: Int = -1,
+                     ref :: Real = - one(eltype(x)),
+                     exp :: Real = - one(eltype(x)) * 2,
+                     ocn :: Real = - one(eltype(x)) / 2,
+                     icn :: Real = one(eltype(x)) / 2)
+
+  # Initial simplex
+  n = nlp.meta.nvar
+  T = eltype(x)
+  if n < length(vertices) < 0
```
Comment (abelsiqueira) [inline src/nelder_mead.jl]: This doesn't look right

File: src/nelder_mead.jl [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+    error("Invalid Simplex : The number of initial points is less than n + 1")
+  elseif length(vertices) == 0
+    vertices = [x,[copy(x) for i = 1:n]...]
```
Comment (abelsiqueira) [inline src/nelder_mead.jl]: Add a space `[x, [`.

File: src/nelder_mead.jl [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+    for j in 1:n
+      xt = vertices[j + 1]
+      xt[j] += xt[j] == 0 ? eps(T)^T(0.25) : T(0.05)
+    end
+  end
+
+  pairs = [[x, obj(nlp, x)] for x in vertices]
+  sort!(pairs, by=x->x[2])
+
+  x_trial = zeros(eltype(x), n)
```
Comment (abelsiqueira) [inline src/nelder_mead.jl]: Use already have `T = eltype(x)`.

File: src/nelder_mead.jl [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+
+  k = 0
+  el_time = 0.0
+  start_time = time()
+  tired = neval_obj(nlp) > max_eval >= 0 || el_time > max_time
+  norm_opt = norm(first(pairs)[1] - last(pairs)[1])
+  optimal = norm_opt < tol
+  T = eltype(x)
+  status =:unknown
+  @info log_header([:iter, :f, :nrm], [Int, T, T], hdr_override=Dict(:f => "f(x)", :nrm => "‖x₁ - xₙ+1‖"))
```
Comment (abelsiqueira) [inline src/nelder_mead.jl]: Use can use `x₁ - xₙ₊₁` (`\_+[TAB]`)

**Comment by abelsiqueira** [comment]
Thanks!
